Source/DafnyLanguageServer.Test/Unit/ParserExceptionTest.cs

```diff
@@ -0,0 +1,99 @@
+﻿using System;
+using Microsoft.Dafny.LanguageServer.Language;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Boogie;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.Dafny.LanguageServer.IntegrationTest.Unit {
+
+  [TestClass]
+  public class ParserExceptionTest {
+    private static readonly string TestFilePath = "parserException.dfy";
+    private const string LanguageId = "dafny";
+    private const int MaxTestExecutionTimeMs = 10000;
+    private DafnyLangParser parser;
+    private LastDebugLogger lastDebugLogger;
+
+    [TestInitialize]
+    public void SetUp() {
+      lastDebugLogger = new LastDebugLogger();
+      parser = DafnyLangParser.Create(lastDebugLogger);
+    }
+
+    [TestMethod, Timeout(MaxTestExecutionTimeMs)]
+    public void DocumentWithParserExceptionDisplaysIt() {
+      var source = "function t() { / }";
+      var documentItem = CreateTestDocument(source, TestFilePath);
+      var errorReporter = new ParserExceptionSimulatingErrorReporter();
+      parser.Parse(documentItem, errorReporter, new CancellationToken());
+      Assert.AreEqual($"encountered an exception while parsing file:///{TestFilePath}", lastDebugLogger.LastDebugMessage);
+      Assert.AreEqual($"/{TestFilePath}(1,0): Error: [internal error] Parser exception: Simulated parser internal error", errorReporter.LastMessage);
+    }
+
+    /// <summary>
+    /// An error reporter that throws an exception on the first reported error, to simulate a parser exception.
+    /// </summary>
+    class ParserExceptionSimulatingErrorReporter : ErrorReporter {
+      private int numberOfErrors;
+      public string LastMessage = "";
+      public override bool Message(MessageSource source, ErrorLevel level, IToken tok, string msg) {
+        numberOfErrors++;
```

Comment:
This statement should probably be inside the if-statement below.
Just out of curiosity: Is it necessary to limit throwing the exception only upon the first error? Wouldn't it be easier to ensure that the input Dafny code only yields one error? Or is this some sort of fail-safety to not break this unit test in the future if the language changes?

Reply:
Yes it is necessary to limit to throwing the exception only upon the first error (the simulated parsing exception), because there is now a try/catch that will emit another error afterwards. (`errorReporter.Error`)